### ISSUE TITLE
stopWorkflow does not work inside NodeJS : ReferenceError: document is not defined

### ISSUE THREAD

**Issue by saunlani** [Aymkdn/SharepointPlus#162]
This is the error I receive when attempting to use stopWorkflow in an external NodeJS API:

my execution:
`sp.list("ListName","https://portal.corp.company.com/").stopWorkflow({ID:6736, workflowName:"Employee Workflow 2010"});`

the error:
```
(node:14288) UnhandledPromiseRejectionWarning: ReferenceError: document is not defined
    at C:\Repos\EmployeeAPI\node_modules\sharepointplus\dist\lists\stopWorkflow.js:50:7
    at new Promise (<anonymous>)
    at new Wrapper (C:\Repos\EmployeeAPI\node_modules\core-js-pure\internals\export.js:15:24)
    at C:\Repos\EmployeeAPI\node_modules\sharepointplus\dist\lists\stopWorkflow.js:48:12
(node:14288) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was
not handled with .catch(). (rejection id: 2)

```

**Comment by Aymkdn** [comment]
There is no API to stop a workflow. So as a workaround the script opens an iframe to the Workflow Instance page where there is the "stop workflow" option. And then the form is posted to cancel the workflow.

In a Node environment, I'm not sure right now. I'll have to do some research to see how it's possible to have the same kind of behavior.

**Comment by saunlani** [comment]
I see, yeah this function is actually my main purpose behind using your package. So any help at all is greatly appreciated!

**Comment by Aymkdn** [comment]
I'll need a few days to find the time to work on it.

**Comment by Aymkdn** [comment]
I don't have the bandwidth right now to investigate. If you cannot wait for a couple of weeks, then I'll invite you to explore by yourself the below solutions:
- https://github.com/jsdom/jsdom – that can interpret HTML/JS from a page
- Puppeter https://pptr.dev/ – that can open a browser (without the UI) and runs through a page

You have to use [`$SP().getWorkflowID()`](https://aymkdn.github.io/SharepointPlus/lists.html#getWorkflowID) to get the `instances`: you need the last instance that contains an url. Something like:
```js
async function run() {
  let params = await $SP().getWorkflowID({ID:itemID, workflowName:"The workflow name"});
  if (params.instances.length===0) return "No instance available for this workflow/item ID.";
  let pageUrl = params.instances[params.instances.length-1].StatusPageUrl;

  // then call "pageUrl" with either `jsdom` or a puppeter script
  // once the page is loaded, you need to trigger the below from the page:
  // window.__doPostBack('ctl00$PlaceHolderMain$HtmlAnchorEnd','')
}
```

**Comment by Aymkdn** [comment]
I took some time to find out a solution. It took me one hour, and here is how to proceed: https://github.com/Aymkdn/SharepointPlus/wiki/Stop-a-workflow-2010-from-Node-Server-Application

**Comment by saunlani** [comment]
seriously, thank you very much for resolving this so quickly!

**Comment by Aymkdn** [comment]
I don't know. I'll have to create a list and a workflow and test.

**Comment by Aymkdn** [comment]
I've just tried, and it worked as expected: I manually started a workflow (that is doing a "pause for 5 hours") then I used the code to stop it, and it worked.

**Comment by Aymkdn** [comment]
It's using the "End this workflow" link from a Workflow Status page. So if you can see this link for the workflow/item you want to stop, then it's supposed to work:
![image](https://user-images.githubusercontent.com/946315/106467338-ca801700-649c-11eb-8678-7e24586ea33d.png)

**Comment by saunlani** [comment]
Interesting.  Can you please try starting two workflows and using a for loop to iterate through them and stop both of them?

**Comment by Aymkdn** [comment]
Sorry, I don't have the time to do it. Please, refer to [my previous comment](https://github.com/Aymkdn/SharepointPlus/issues/162#issuecomment-770871102).

**Comment by saunlani** [comment]
Yeah I've tested this thoroughly and it seems like the stopWorkflow function is only working for the very first workflow.  It has no effect on other workflows.

**Comment by Aymkdn** [comment]
It seems there is an issue in my code as the `getWorkflowID` will only return the data for the first workflow. I'll have to fix it.

**Comment by Aymkdn** [comment]
I've just published v6.1.4 that fixes `$SP().list().getWorkflowID()` and also that makes `$SP().list().stopWorkflow()` works in Node app (so no need to use the wiki page that I'm going to delete)

**Comment by saunlani** [comment]
Thanks again sir!
